### PR TITLE
export get pinnable slice

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**

### CODE REVIEW

File: src/rocksdb.rs [hunk 1]
```diff
@@ -1306,8 +1352,14 @@ impl Deref for DBVector {
...
     fn drop(&mut self) {
-        unsafe {
-            libc::free(self.base as *mut c_void);
+        if self.is_pinned {
+            unsafe {
+                crocksdb_ffi::crocksdb_pinnableslice_destroy(self.pinned_slice);
+            }
+        } else {
+            unsafe {
+                libc::free(self.base as *mut c_void);
```
Comment (siddontang) [inline src/rocksdb.rs]: do we need this now? I think all can use pinnable slice.

File: librocksdb_sys/src/lib.rs [hunk 2]
```diff
@@ -764,6 +765,22 @@ extern "C" {
...
+                               err: *mut *mut c_char)
+                               -> *mut DBPinnableSlice;
+    pub fn crocksdb_get_pinned_cf(db: *mut DBInstance,
+                                  readopts: *const DBReadOptions,
+                                  cf_handle: *mut DBCFHandle,
+                                  k: *const u8,
+                                  kLen: size_t,
+                                  err: *mut *mut c_char)
+                                  -> *mut DBPinnableSlice;
+    pub fn crocksdb_pinnableslice_value(s: *mut DBPinnableSlice, valLen: *const size_t) -> *mut u8;
```
Comment (BusyJay) [inline librocksdb_sys/src/lib.rs]: Why const?

Comment (andelf) [inline librocksdb_sys/src/lib.rs]: `->*mut u8` should be `*const u8`

**Comment by andelf** [comment]
IMHO, embedding DBPinnedSlice into DBVector is a bad idea, leaves multi_get #59 difficult to implement.
Better distinguish PinnedSlice and old DBVector(which allocates mem in crocksdb code).

Or totally rewrite old DBVector as a wrapper of *std::string

**Comment by siddontang** [comment]
I think we can close multi-get PR now, it has no effective and hard to use. And I think RocksDB will support pinnable multi get if many users want this.

**Comment by zhangjinpeng1987** [comment]
The old API that accepts std::string, is discouraged.

**Comment by siddontang** [comment]
PTAL @BusyJay @andelf

File: librocksdb_sys/src/lib.rs [hunk 3]
```diff
@@ -764,6 +765,22 @@ extern "C" {
...
+                               err: *mut *mut c_char)
+                               -> *mut DBPinnableSlice;
+    pub fn crocksdb_get_pinned_cf(db: *mut DBInstance,
+                                  readopts: *const DBReadOptions,
+                                  cf_handle: *mut DBCFHandle,
+                                  k: *const u8,
+                                  kLen: size_t,
+                                  err: *mut *mut c_char)
+                                  -> *mut DBPinnableSlice;
+    pub fn crocksdb_pinnableslice_value(s: *mut DBPinnableSlice, valLen: *mut size_t) -> *mut u8;
```
Comment (BusyJay) [inline librocksdb_sys/src/lib.rs]: I think it should be `*const DBPinnableSlice` and `*const u8`.

File: src/rocksdb.rs [hunk 4]
```diff
@@ -1272,6 +1266,7 @@ impl Writable for WriteBatch {
 pub struct DBVector {
     base: *mut u8,
     len: usize,
+    pinned_slice: *mut DBPinnableSlice,
```
Comment (BusyJay) [inline src/rocksdb.rs]: `base` can be removed.

File: librocksdb_sys/crocksdb/c.cc [hunk 5]
```diff
@@ -2853,4 +2855,49 @@ crocksdb_logger_t *crocksdb_create_log_from_options(const char *path,
...
+    return NULL;
+  }
+  return v;
+}
+
+void crocksdb_pinnableslice_destroy(crocksdb_pinnableslice_t* v) { delete v; }
+
+const char* crocksdb_pinnableslice_value(const crocksdb_pinnableslice_t* v,
+                                        size_t* vlen) {
+  if (!v) {
```
Comment (BusyJay) [inline librocksdb_sys/crocksdb/c.cc]: The check if redundant.

Comment (siddontang) [inline librocksdb_sys/crocksdb/c.cc]: if v is null, no check will cause panic.

Comment (BusyJay) [inline librocksdb_sys/crocksdb/c.cc]: But it can't be null.

**Comment by andelf** [comment]
LGTM

File: librocksdb_sys/crocksdb/c.cc (hunk 5 above)
Comment (zhangjinpeng1987) [inline librocksdb_sys/crocksdb/c.cc]: In our case, when call `crocksdb_pinnableslice_value` the `v` is not null, but there is no guarantee that `v` is not null in other cases.

File: src/rocksdb.rs [hunk 6]
```diff
@@ -1270,54 +1264,56 @@ impl Writable for WriteBatch {
...
 
 pub struct DBVector {
-    base: *mut u8,
-    len: usize,
+    pinned_slice: *mut DBPinnableSlice,
 }
 
 impl Debug for DBVector {
     fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        let mut val_len: size_t = 0;
```
Comment (BusyJay) [inline src/rocksdb.rs]: `write!(formatter, "{:?}", &**self)` should be enough.

File: src/rocksdb.rs [hunk 7]
```diff
@@ -1270,54 +1264,56 @@ impl Writable for WriteBatch {
...
+            let val = crocksdb_ffi::crocksdb_pinnableslice_value(self.pinned_slice, val_len_ptr);
+            write!(formatter, "{:?}", slice::from_raw_parts(val, val_len))
         }
     }
 }
 
 impl<'a> PartialEq<&'a [u8]> for DBVector {
     fn eq(&self, rhs: &&[u8]) -> bool {
-        if self.len != rhs.len() {
+        let mut val_len: size_t = 0;
```
Comment (BusyJay) [inline src/rocksdb.rs]: `**rhs == **self` should be enough.

File: librocksdb_sys/crocksdb/c.cc (hunk 5 above)
Comment (BusyJay) [inline librocksdb_sys/crocksdb/c.cc]: What are the other cases?

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**